extensions/amp-subscriptions/0.1/amp-subscriptions.js

```diff
@@ -301,41 +300,71 @@ export class SubscriptionService {
         return this.platformStore_.resolveEntitlement('local',
             Entitlement.empty('local'));
       }
-      const decodedData = this.jwtHelper_.decode(authData);
-      user().assert(
-          decodedData['aud'] == origin ||
-          decodedData['aud'] == sourceOrigin,
+
+      return this.verifyAuthToken_(authData).then(entitlement => {
+        // Viewer authorization is redirected to use local platform instead.
+        this.platformStore_.resolveEntitlement('local', entitlement);
+      }).catch(reason => this.sendAuthTokenErrorToViewer_(reason));
+
+    }, reason => {
+      throw user().createError('Viewer authorization failed', reason);
+    });
+  }
+
+  /**
+   * Logs error and sends message to viewer
+   * @param {string} token
+   * @returns {!Promise<!Entitlement>}
+   */
+  verifyAuthToken_(token) {
+    const origin = getWinOrigin(this.ampdoc_.win);
+    const sourceOrigin = getSourceOrigin(this.ampdoc_.win.location);
+    const decodedData = this.jwtHelper_.decode(token);
+    const currentProductId = /** @type {string} */ (user().assert(
+        this.pageConfig_.getProductId(),
+        'Product id is null'
+    ));
+    if (decodedData['aud'] != origin || decodedData['aud'] == sourceOrigin) {
+      return Promise.reject(
```

Comment:
We currently do not allow `Promise.reject()` w/o `Error` object. One easy path is to keep `user().assert()`, but wrap the whole body into `new Promise(resolve => {...})`. Then these errors will be automatically rejected. Otherwise, `user().createError()` would also work.